### DISCUSSION TITLE
Grid view

### DISCUSSION THREAD

**Discussion by whizzer929** [Azure/ipam#135] [Ideas]
Hello Ipam Team,

Just submitting this as a potential feature request.  Would like a more traditional IPAM graphical grid view, not so much for utilization, but more of a geographical view of where usage is.  Attached are a couple of examples.  This could be done on Space or Block level?  And maybe an option to change the scale.  I.E. each block = /27, /24, etc.

Thanks!
![ipam grid](https://github.com/Azure/ipam/assets/65122209/b12454b4-6d2f-4a1f-8a90-cd6a403e3e18)
![ipam grid 2](https://github.com/Azure/ipam/assets/65122209/54499f27-c76f-4720-8481-6aea74f21499)

**Comment by DCMattyG** [comment]
Hey @whizzer929, I'm familiar with this type of view from Infoblox. I know it's not exactly the same, but have you seen the Planner feature yet?

![image](https://github.com/Azure/ipam/assets/12955732/1f0d0194-3b3f-4323-ba2a-0d0f4f298d7b)

I'm guessing this is not sufficient enough for what you're trying to do with the grid view?

It does support view by Space/Block, or by simply Virtual Network (for those Virtual Networks not managed by IPAM already). You can change the Subnet Mask scope, and it will even indicate where overlap exists (red areas).

**Reply by whizzer929** [reply]
Hey DCMattyG,  yes we use the planner and it works well for VNets, but doesn't work with Spaces/Blocks as you mentioned.  This is the idea, but maybe in a more condensed format like in the attachments.  Perhaps that python module has some options for this.

**Reply by DCMattyG** [reply]
Got it, I think what you're saying is that it doesn't work at the Space/Block **level** (meaning multiple Virtual Networks contained within).

Would you like this delivered as some sort of graphical representation within the UI, or would you like it as an Excel spreadsheet that you can download for a view similar to what you have above (ideally)?

**Reply by whizzer929** [reply]
Definitely in the UI.  Downloading a spreadsheet is probably a nice to have, but not necessary, for me anyway.

**Reply by DCMattyG** [reply]
Makes sense. And final question....

Do you need any sort of interaction with the graphic once it's generated (clicking, etc.), or just a way to generate a visual representation of the utilization at a Space/Block level would be good enough for what you need?

**Reply by whizzer929** [reply]
No, I don't need any interaction, a static picture would be fine.  Thanks again!

**Reply by DCMattyG** [reply]
Ok, I've added this to my plan. I'm working on some critical features during July, so will probably look into this in August if that's alright with you.

**Reply by whizzer929** [reply]
OK.  Just thinking though, a scale option might be useful.  I.E. 1 square grid block = /28, /26, /21, etc.  And this would add or subtract gird blocks based on the source Space/Block size.  A kind of zone in/out option?

**Reply by DCMattyG** [reply]
I see, so each "block" in the grid would be of a defined size from something like a drop-down. Max size would be the CIDR mask of the Block itself?

This makes sense when looking at the Block view, but at the Space level you could have Blocks all over the place and no efficient way to "summarize" the view.....thoughts on that?

**Reply by whizzer929** [reply]
This makes sense when looking at the Block view, but at the Space level you could have Blocks all over the place and no efficient way to "summarize" the view.....thoughts on that?

Assuming you can't add overlapping blocks to a space, I've never tried that, and all blocks would reside in the space, maybe something like gray out the IP Space not covered by blocks.  Something like this?

![ipam grid 3](https://github.com/Azure/ipam/assets/65122209/06670c3f-c5e4-4b07-b5ae-b128e8ba2954)

**Reply by DCMattyG** [reply]
That makes sense, and you **cannot** add overlapping Blocks to a Space. A Space, by definition, is a collection of *unique* IP space/addresses (e.g. non-overlapping).

I like what you've laid out above. For the "Block" view, the zoom-out threshold would be the CIDR Mask of the Block, but I guess in the Space context the zoom-out threshold would just be "whatever" (or perhaps max of /8). Does that make sense to you?

**Reply by whizzer929** [reply]
For the Space zoom out max, yeah /8 should be fine.  Which is the mask for the RFC-1918-Private Ranges Class A.

**Reply by DCMattyG** [reply]
Excellent, notes added. Will circle back in August 👍
